GreedyParameterPointSelector.C

```diff
@@ -0,0 +1,465 @@
+                    dist_to_points[j] += dist;
+                }
+            }
+
+            double closest_dist_to_points = INT_MAX;
+            for (int i = 0; i < dist_to_points.size(); i++) {
+                if (dist_to_points[i] < closest_dist_to_points)
+                {
+                    closest_dist_to_points = dist_to_points[i];
+                    d_next_point_to_sample = i;
+                }
+            }
+        }
+    }
+
+    d_iteration_started = true;
+    d_max_error = 0;
+
+    int curr_point_to_sample = d_next_point_to_sample;
+    d_next_point_to_sample = -1;
+
+    d_parameter_point_errors[curr_point_to_sample] = 0;
+
+    d_subset_created = false;
+    d_subset_counter = 0;
+    d_counter = -1;
+
+    auto search = d_parameter_sampled_indices.find(curr_point_to_sample);
+    CAROM_VERIFY(search == d_parameter_sampled_indices.end());
+
+    d_parameter_sampled_indices.insert(curr_point_to_sample);
+
+    return curr_point_to_sample;
+}
+
+int
+GreedyParameterPointSelector::getNextPointRequiringResidual()
+{
+    if (d_parameter_sampled_indices.size() == d_parameter_points.size())
+    {
+        d_procedure_completed = true;
+        return -1;
+    }
+    if (d_subset_counter == d_subset_size)
+    {
+        return -1;
+    }
+    if (!d_iteration_started || d_procedure_completed)
+    {
+        return -1;
+    }
+    if (d_point_requiring_residual_computed)
+    {
+        return -1;
+    }
+
+    if(!d_subset_created)
+    {
+        // generate random shuffle
+        if (!d_debug_algorithm)
+        {
+          std::shuffle(d_parameter_point_random_indices.begin(), d_parameter_point_random_indices.end(), rng);
+        }
+        d_subset_created = true;
+    }
+
+    d_next_point_requiring_residual = -1;
+
+    //get next point requiring residual
+    while (d_counter < (int) d_parameter_points.size() - 1)
+    {
+        d_counter++;
+        if (d_subset_counter == d_subset_size)
+        {
+            break;
+        }
+        auto search = d_parameter_sampled_indices.find(d_parameter_point_random_indices[d_counter]);
+        if (search == d_parameter_sampled_indices.end())
+        {
+            d_subset_counter++;
+            if (d_sat * d_parameter_point_errors[d_parameter_point_random_indices[d_counter]] > d_max_error)
+            {
+                d_next_point_requiring_residual = d_parameter_point_random_indices[d_counter];
+                d_point_requiring_residual_computed = true;
+                return d_next_point_requiring_residual;
+            }
+        }
+    }
+
+    if (d_next_point_requiring_residual == -1)
+    {
+        d_iteration_started = false;
+    }
+
+    return d_next_point_requiring_residual;
+}
+
+void
+GreedyParameterPointSelector::setPointResidual(double error, int rank, int num_procs)
+{
+    CAROM_VERIFY(error >= 0);
+    CAROM_VERIFY(d_point_requiring_residual_computed);
+
+    double *proc_errors = new double[num_procs];
```

Comment:
Instead of allocating this array and gathering, you can just do `MPI_Allreduce` on the scalar local error squared.